File: daily_problems/2025/04/0404/personal_submission/cf1530d_nhantran0.cpp
/**
 * https://codeforces.com/problemset/problem/1530/D
 * D. Secret Santa
 * Solver: nhant
 * 2025-04-04 22:48:21
 */
#include <bits/stdc++.h>

using namespace std;

void solve() {
    int n;
    cin >> n;
    vector<int> b(n + 1);
    vector<vector<int>> from(n + 1, vector<int>());
    for (int i = 1; i <= n; i++) {
        cin >> b[i];
        from[b[i]].push_back(i);
    }
    int k = n;
    for (int i = 1; i <= n; i++) {
        if (from[i].size() > 1) { // in deg > 1
            int sawKeep = false;
            for (int j = 0, m = (int)from[i].size(); j < m; j++) {
                int u = from[i][j];
                // prioritize breaking edges of a "path" to create at least 1 path (from a circle) or 2 paths (from a path)
                if (!sawKeep && (from[u].size() == 0 || j == m - 1)) {
                    sawKeep = true;
                } else {
                    b[u] = 0;
                    k--;
                }
            }
        }
    }
    vector<pair<int,int>> paths;
    for (int i = 1; i <= n; i++) {
        if (from[i].size() == 0) {
            int end = i;
            while (b[end] != 0) end = b[end];
            // BEWARE that this causes inf loop. Find the test case input!
            // if (paths.size() > 0) {
            //     b[paths[paths.size() - 1].second] = i;
            // }
            paths.push_back({i, end});
        }
    }
    for (int i = 0, z = (int) paths.size(); i < z; i++) {
        int j = i == z - 1 ? 0 : i + 1;
        b[paths[i].second] = paths[j].first;
    }
    cout << k << '\n';
    for (int i = 1; i <= n; i++) {
        cout << b[i] << " ";
    }
    cout << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    int tt;
    cin >> tt;
    while (tt--) {
        solve();
    }
}

File: daily_problems/2025/04/0404/personal_submission/cf1530d_nhantran0.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

/**
 * Refs: <a href="https://codeforces.com/problemset/problem/1530/D">D. Secret Santa</a>
 * Author: nhant
 * 2025-04-04 21:47:45 +0800
 **/
public class D {
  static IO io;

  private static void solve() {
    int n = io.nextInt();
    int[] b = new int[n + 1];
    List<List<Integer>> from = new ArrayList<>(n + 1);
    for (int i = 0; i <= n; i++) from.add(new ArrayList<>());
    for (int i = 1; i <= n; i++) {
      b[i] = io.nextInt();
      from.get(b[i]).add(i);
    }
    int k = n;
    for (int i = 1; i <= n; i++) {
      if (from.get(i).size() > 1) {
        List<Integer> vts = from.get(i);
        boolean sawKeep = false;
        for (int j = 0, m = vts.size(); j < m; j++) {
          int u = vts.get(j);
          // prioritize breaking edges of a "path" to create at least 1 path (from a circle) or 2 paths (from a path)
          if (!sawKeep && (from.get(u).isEmpty() || j == m - 1)) {
            sawKeep = true;
          } else {
            b[u] = 0;
            k--;
          }
        }
      }
    }
    List<int[]> paths = new ArrayList<>();
    for (int i = 1; i <= n; i++) {
      if (from.get(i).isEmpty()) {
        int end = i;
        while (b[end] != 0) end = b[end];
        paths.add(new int[]{i, end});
      }
    }
    for (int i = 0; i < paths.size(); i++) {
      int j = i == paths.size() - 1 ? 0 : i + 1;
      b[paths.get(i)[1]] = paths.get(j)[0];
    }
    io.println(k);
    for (int i = 1; i <= n; i++)
      io.printf("%d ", b[i]);
    io.println();
  }

  public static void main(String[] args) {
    try {
      io = new IO();
      int tc = io.nextInt();
      while (tc-- > 0) {
        solve();
      }
    } finally {
      io.close();
    }
  }

  static class IO extends PrintWriter {
    private BufferedReader r;
    private StringTokenizer st;

    public IO() {
      this(System.in, System.out);
    }

    public IO(InputStream i, OutputStream o) {
      super(o);
      r = new BufferedReader(new InputStreamReader(i), 1 << 15);
    }

    public String next() {
      try {
        while (st == null || !st.hasMoreTokens()) st = new StringTokenizer(r.readLine());
        return st.nextToken();
      } catch (Exception e) {
      }
      return null;
    }

    public int nextInt() {
      return Integer.parseInt(next());
    }
  }
}
